### PR TITLE
Upgrade3xy

### PR DESCRIPTION
**Pull request by Dee-Fuse**
Some migrations to CakePHP 3.x

### CODE REVIEW

File: src/Model/Table/QueuedTasksTable.php [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 use Cake\Core\Configure;
 use Cake\Utility\Hash;
-use Tools\Model\Table\Table;
+use Cake\ORM\Table;
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: that was on purpose, to use the shim functionality and the "truncate" part :)

Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: So, i need "cakephp-shim" installed to use the Queue?
I have replaced the shim parts with CakePHP default ones to keep my project small

Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: Not necessarily, I was just starting to upgrade, and this kept me from having to do all at once.
I bet we can remove it at some point.
There are still 5-6 failing tests we should take care of first, before optimizing :)

File: src/Model/Table/QueuedTasksTable.php [hunk 2]
```diff
@@ -67,11 +67,11 @@ public function createJob($jobName, $data = null, $notBefore = null, $group = nu
 		$data = [
 			'jobtype' => $jobName,
 			'data' => serialize($data),
-			'group' => $group,
+			'task_group' => $group,
 			'reference' => $reference
 		];
 		if ($notBefore !== null) {
-			$data['notbefore'] = date('Y-m-d H:i:s', strtotime($notBefore));
+			$data['notbefore'] =  strtotime($notBefore);
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: why is this necessary? does the ORM better work with int instead of date strings? maybe we should use Time() object rather?

File: src/Model/Table/QueuedTasksTable.php (hunk 1 above)
Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: Ok, I'll take a look at passing the tests without the shim functions tomorrow.
But the 'group' field should be changed nevertheless ;)

Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: Why? I use the `'quoteIdentifiers' => true` by default :-)

But I guess it might be useful for some people to not have to depend on it.

File: src/Model/Table/QueuedTasksTable.php (hunk 2 above)
Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: The first days with CakePHP 3 made me some problems with saving formatet DATETIME fields when using German formating/language because of the different order 'd.m.Y H:i:s'. The ORM converts the timestamp automatical to an Time object when saving/retrieving.

Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: good to know

File: src/Model/Table/QueuedTasksTable.php (hunk 1 above)
Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: I'm developing into an clean CakePHP installation without any special settings expecting the MySQL stuff.
>  It should be noted that this decreases performance...
and phpMyAdmin complains about it, too.

Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: Performance (especially that kind of) of the queue plugin is irrelevant ;)
but yeah, for the sake of "clean defaults", why not.
In that case we need a DB migration, though.

Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: I'll bake one out of my test DB and Push it asap.

Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: Well, you just need a second one on top of the existing one: https://github.com/dereuromark/cakephp-queue/blob/3.0/config/Migrations/20150425180802_init.php

File: config/Migrations/20150511062806_initial.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * Migrate Down.
+     */
+    public function down()
+    {
+
+    }
+}
```
Comment (dereuromark) [inline config/Migrations/20150511062806_initial.php]: Please make a new file with only the change you need, this first iniial one should not be touched

File: src/Model/Table/QueuedTasksTable.php [hunk 4]
```diff
@@ -286,13 +286,14 @@ public function updateProgress($id, $progress) {
...
-			'completed' => "'" . date('Y-m-d H:i:s') . "'"
-		];
-		$conditions = [
-			'id' => $id
-		];
-		return $this->updateAll($fields, $conditions);
+		$data=$this->get($id);
+		$data->completed = time();
+		
+		if($this->save($data))
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: cs issues

File: src/Model/Table/QueuedTasksTable.php [hunk 5]
```diff
@@ -323,15 +324,12 @@ public function reset() {
...
-		$db = $this->getDataSource();
-		$fields = [
-			'failed' => 'failed + 1',
-			'failure_message' => $db->value($failureMessage),
-		];
-		$conditions = [
-			'id' => $id
-		];
-		return $this->updateAll($fields, $conditions);
+		$data=$this->get($id);
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: also cs

File: src/Model/Table/QueuedTasksTable.php [hunk 6]
```diff
@@ -323,15 +324,12 @@ public function reset() {
...
+
+		$data->failed++;
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: there was a reason why this used atomic operations, your version is not fail-proof on race conditions

File: src/Model/Table/QueuedTasksTable.php [hunk 7]
```diff
@@ -323,15 +324,12 @@ public function reset() {
...
+		$data->failure_message = $failureMessage;
+		
+		return $this->save($data);
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: the return value is now different it should continue using updateAll()

File: src/Shell/Task/QueueRetryExampleTask.php [hunk 8]
```diff
@@ -105,6 +105,7 @@ public function run($data, $id = null) {
 
 		if ($count < 3) {
 			$count++;
+			$this->failureMessage = 'Failed, because '.$count.' < 3';
```
Comment (dereuromark) [inline src/Shell/Task/QueueRetryExampleTask.php]: cs

File: tests/Fixture/QueuedTasksFixture.php [hunk 9]
```diff
@@ -1,33 +1,50 @@
...
-		'failed' => ['type' => 'integer', 'null' => false, 'default' => '0', 'length' => 3],
-		'failure_message' => ['type' => 'text', 'null' => true, 'default' => null],
-		'workerkey' => ['type' => 'string', 'null' => true, 'length' => 45],
-		'_constraints' => ['primary' => ['type' => 'primary', 'columns' => ['id']]]
-	];
+/**
+ * QueuedTasksFixture
+ *
+ */
+class QueuedTasksFixture extends TestFixture
```
Comment (dereuromark) [inline tests/Fixture/QueuedTasksFixture.php]: can we please keep the previous cs?

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 10]
```diff
@@ -1,48 +1,48 @@
 <?php
-/**
- * @author MGriesbach@gmail.com
- * @license http://www.opensource.org/licenses/mit-license.php The MIT License
- * @link http://github.com/MSeven/cakephp_queue
- */
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: we shouldnt remove the copyrights

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 11]
```diff
@@ -1,48 +1,48 @@
...
+     *
+     * @return void
+     */
+    public function setUp()
+    {
+        parent::setUp();
 
-		$this->QueuedTasks = TableRegistry::get('QueuedTasks');
-	}
+		$config = TableRegistry::exists('QueuedTasks') ? [] : ['className' => 'Queue\Model\Table\QueuedTasksTable'];
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: why is that necessary, there is always the class.

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 12]
```diff
@@ -422,56 +423,56 @@ public function testRequestGroup() {
...
 		// create an ungrouped task
-		$this->assertTrue((bool)$this->QueuedTasks->createJob('task1', 1));
+		$this->assertTrue((bool)$this->QueuedTasksTable->createJob('task1', 1));
 		//create a Grouped Task
-		$this->assertTrue((bool)$this->QueuedTasks->createJob('task1', 2, null, 'testgroup'));
+		$this->assertTrue((bool)$this->QueuedTasksTable->createJob('task1', 2, null, 'testgroup'));
 
 		// Fetching without group should completely ignore the Group field.
-		$this->QueuedTasks->clearKey();
-		$tmp = $this->QueuedTasks->requestJob($capabilities);
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: Changing all those was not necessary and only makes the diff harder to read. Please try to stick to the absolute minimum of changes.

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 13]
```diff
@@ -422,56 +423,56 @@ public function testRequestGroup() {
...
-		$this->QueuedTasks->clearKey();
-		$tmp = $this->QueuedTasks->requestJob($capabilities, 'testgroup');
+		$this->QueuedTasksTable->clearKey();
+		$tmp = $this->QueuedTasksTable->requestJob($capabilities, 'testgroup');
 		$this->assertEquals($tmp['jobtype'], 'task1');
 		$this->assertEquals(6, unserialize($tmp['data']));
 
 		// use FindProgress on the testgroup:
-		$progress = $this->QueuedTasks->find('progress', [
+		$progress = $this->QueuedTasksTable->find('all', [
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: do we stil have the custom finder available? sounds useful.

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 11 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: Just a suggestion from bake

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 10 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: Oh, that was a copy/paste misstake because i've created the Test again with the "bake" command

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 13 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: Not without the Tools installed

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 12 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: I'll keep it in mind the next time

File: src/Model/Table/QueuedTasksTable.php (hunk 6 above)
Comment (Dee-Fuse) [inline src/Model/Table/QueuedTasksTable.php]: by default the function `$this->getDataSource();` has been replaced in CakePHP 3

File: config/Migrations/20150511062806_fixmissing.php [hunk 14]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+use Phinx\Migration\AbstractMigration;
+
+class Initial extends AbstractMigration {
```
Comment (dereuromark) [inline config/Migrations/20150511062806_fixmissing.php]: the class needs to be called differently afaik

File: src/Model/Table/QueuedTasksTable.php [hunk 15]
```diff
@@ -323,15 +323,18 @@ public function reset() {
 	 * @return bool Success
 	 */
 	public function markJobFailed($id, $failureMessage = null) {
-		$db = $this->getDataSource();
+		$db = $this->get($id);
+		if(is_null($failureMessage)) {
```
Comment (dereuromark) [inline src/Model/Table/QueuedTasksTable.php]: === null - and cs issue

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 11 above)
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: For our tests we dont need that afaik.

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 16]
```diff
@@ -458,20 +464,20 @@ public function testRequestGroup() {
...
 
 		$this->assertNull($progress[0]['reference']);
-		$this->assertEquals($progress[0]['status'], 'IN_PROGRESS');
+#		$this->assertEquals($progress[0]['status'], 'IN_PROGRESS');
 		$this->assertEquals($progress[1]['reference'], 'Job number 4');
-		$this->assertEquals($progress[1]['status'], 'IN_PROGRESS');
+#		$this->assertEquals($progress[1]['status'], 'IN_PROGRESS');
 		$this->assertEquals($progress[2]['reference'], 'Job number 6');
-		$this->assertEquals($progress[2]['status'], 'IN_PROGRESS');
+#		$this->assertEquals($progress[2]['status'], 'IN_PROGRESS');
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: those tests dont work anymore?

**Comment by dereuromark** [comment]
Looking good so far!

File: config/Migrations/20150511062806_fixmissing.php (hunk 14 above)
Comment (Dee-Fuse) [inline config/Migrations/20150511062806_fixmissing.php]: Oops, missed to rename the class, will be fixed

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 16 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: nop, the field status will never be filled while executing, I've added the status field with my additional migration that it exists in the database

File: config/Migrations/20150511062806_fixmissing.php [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+	 *
+	 * Uncomment this method if you would like to use it.
+	 *
+	 * @return void
+	 */
+
+	public function change() {
+		$table = $this->table('queued_tasks');
+		$table->addColumn('status', 'string', array('limit' => 255))
+			->create();
```
Comment (dereuromark) [inline config/Migrations/20150511062806_fixmissing.php]: where is the group rename? :)

File: src/Shell/Task/QueueLongExampleTask.php [hunk 18]
```diff
@@ -75,7 +75,7 @@ public function add() {
 	 * @param array $data The array passed to QueuedTask->createJob()
 	 * @param int $id The id of the QueuedTask
 	 * @return bool Success
-	 * @throws RuntimeException when seconds are 0;
+	 * @throws Queue\Shell\Task\RuntimeException when seconds are 0;
```
Comment (dereuromark) [inline src/Shell/Task/QueueLongExampleTask.php]: I think its \RuntimeException

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 19]
```diff
@@ -9,45 +9,58 @@
...
-	 * @var array
-	 */
+/**
+ * Fixtures
+ *
+ * @var array
+ */
 	public $fixtures = [
-		'plugin.Queue.QueuedTasks'
+		'plugin.queue.queued_tasks'
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: the previous was correct.

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 20]
```diff
@@ -9,45 +9,58 @@
...
+ * @return void
+ */
 	public function setUp() {
 		parent::setUp();
 
-		Configure::write('App.namespace', 'TestApp');
-
-		$this->QueuedTasks = TableRegistry::get('QueuedTasks');
+		$config = TableRegistry::exists('QueuedTasks') ? [] : ['className' => 'Queue\Model\Table\QueuedTasksTable'];
+		$this->QueuedTasks = TableRegistry::get('QueuedTasks', $config);
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: we can keep the former code, including the namespace!

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php [hunk 21]
```diff
@@ -9,45 +9,58 @@
...
 	}
 
-	/**
-	 * Test the basic create and length evaluation functions.
-	 */
+/**
+ * Test the basic create and length evaluation functions.
+ *
+ * @return void
+ */
```
Comment (dereuromark) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: psr2-r means the indentation was correct before (same level of indentation). dont care about the cs travis report here, that one is not adjusted just yet for this.

File: src/Shell/Task/QueueLongExampleTask.php (hunk 18 above)
Comment (Dee-Fuse) [inline src/Shell/Task/QueueLongExampleTask.php]: phpcs said it should be `Queue\Shell\Task\RuntimeException`

File: tests/TestCase/Model/Table/QueuedTasksTableTest.php (hunk 20 above)
Comment (Dee-Fuse) [inline tests/TestCase/Model/Table/QueuedTasksTableTest.php]: I've moved it th `initialize()` 
I'll change it back and test it without that change

File: src/Shell/Task/QueueLongExampleTask.php (hunk 18 above)
Comment (dereuromark) [inline src/Shell/Task/QueueLongExampleTask.php]: unlikely, only because you didnt put it into the namespace
but there is no such file anyway, so using \ prefix is sufficient.

**Comment by dereuromark** [comment]
Thank you very much

**Comment by Dee-Fuse** [comment]
You're welcome
